### PR TITLE
2991 invalid description query error response

### LINKED ISSUE #2991: Invalid description in POST /op/query error response
Orion 1.8.0-next (31f6c9692eccc7a0d38ba639bc8c4301e40db74a)

Request (metadata is wrong, it should be an array):

```
POST /v2/op/query
{"entities": [{"idPattern": ".*", "typePattern": ".*"}], "metadata": "*", "attributes": []}
```

Response:

```
{"error":"BadRequest","description":"the field /attributes/ must be a JSON array"}
```

Bug: it should be `/metadata/` in the error message, not `/attribute/`

### PR DESCRIPTION
**Pull request by arigliano**
This PR fixes #2991 issue.

- It was implemented a new **parseStringList** method, used in **parseBatchQuery**, in order to correctly parse the string list in **metadata** field of the batch query and then correctly print out the error message, in case of wrong value in this field.

### CODE REVIEW

**Comment by kzangeli** [comment]
About the 
```
FIXME P2: Change name of AttributeList to StringVector/StringList
```

The idea was to change the name of AttributeList to StringList, a more neutral name now that both attributes and metadata use the class (when AttributeList was first created, only attributes used it).

From what I can see, instead of changing the name of the files/classes, a new copy has been created.
If this is correct, then this is no good. We don't need a new copy, we just want to rename the existing one

File: test/functionalTest/cases/2991_invalid_description_query_error_response/invalid_description_query_error_response.test [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. Attempt to perform a batch query with invalid metadata field, and see it fail with a proper error description
+#
+
+echo "01. Attempt to perform a batch query with invalid metadata field, and see it fail with a proper error description"
+echo "============================================================================================="
```
Comment (fgalan) [inline test/functionalTest/cases/2991_invalid_description_query_error_response/invalid_description_query_error_response.test]: Minor: as many "=" as chars in the echo message previous to this line.

**Comment by fgalan** [comment]
I agree with @kzangeli . Duplicating code is not a good idea in general. In fact, I could be missing something but I'm not see the relation between solving that FIXME P2 in the code and the fix for #2991.... In other words, can #2991 issue be solved easily *without needing to address* the FIXME P2 issue?

**Comment by arigliano** [comment]
I agree with @kzangeli for the code duplication, i misundersood. If i've understood correctly, the final idea is to change the AttributeList and the related parseAttributeList to StringList, and not to replicate. This data structure and its parsing method, should handle generic String Lists, such as those coming from either "attribute" or "metadata" fields. The FIXME P2 would be related to this fix because the parsing method (parseAttributeList or the possible parseStringList) should accept the name of the field from which the String List arrives, in order to correctly emit the appropriate error string. In fact, in the current parseAttributeList implementation, the error string **"the field /attributes/ must be a JSON array"** is hardcoded, instead it should depend on the input field name that causes the error (attribute and metadata at this moment). I would propose for sure to remove the code duplication and then to rename all to StringList and parseStringList, and to modify the parsing method, by adding the **fieldName** parameter  as I had done in the first commit.

**Comment by kzangeli** [comment]
That's it :)

File: test/functionalTest/cases/2991_invalid_description_query_error_response/invalid_description_query_error_response.test (hunk 1 above)
Comment (arigliano) [reply test/functionalTest/cases/2991_invalid_description_query_error_response/invalid_description_query_error_response.test]: Fixed in 9f260d5

File: doc/manuals/user/docker.md [hunk 2]
```diff
@@ -1 +0,0 @@
-../../../docker/README.md
```
Comment (fgalan) [inline doc/manuals/user/docker.md]: Diff shows two opposite modifications on the same file but in different diff boxes... Weird, although I understand that without effect.

File: src/lib/jsonParseV2/parseBatchQuery.cpp [hunk 3]
```diff
@@ -103,7 +103,7 @@ std::string parseBatchQuery(ConnectionInfo* ciP, BatchQuery* bqrP)
     }
     else if (name == "attributes")
     {
-      std::string r = parseAttributeList(ciP, iter, &bqrP->attributeV);
+      std::string r = parseStringList(ciP, iter, &bqrP->attributeV,name);
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseBatchQuery.cpp]: Style: whitespace after last comma.

File: src/lib/ngsi/StringList.cpp [hunk 4]
```diff
@@ -65,20 +65,20 @@ void AttributeList::fill(const std::string& commaSeparatedList)
...
   std::string  out = "";
 
-  if (attributeV.size() == 0)
+  if (stringV.size() == 0)
   {
     return "";
   }
 
-  out += startTag("attributes", true);
+  out += startTag("strings", true);
```
Comment (fgalan) [inline src/lib/ngsi/StringList.cpp]: Not sure about this change... Whoi is using this render() method? The NGSIv1 responses rendering? Has you tested full functional test regression after the changes?

File: src/lib/ngsi/StringList.cpp [hunk 5]
```diff
@@ -170,23 +170,23 @@ void AttributeList::push_back(const std::string& attributeName)
...
 * push_back_if_absent - 
 */
-void AttributeList::push_back_if_absent(const std::string& attributeName)
+void StringList::push_back_if_absent(const std::string& string)
 {
-  if (lookup(attributeName) == false)
+  if (lookup(string) == false)
   {
-    attributeV.push_back(attributeName);
+	  stringV.push_back(string);
```
Comment (fgalan) [inline src/lib/ngsi/StringList.cpp]: Style: excessive whitespace (it should be 2 spaces)

File: src/lib/ngsi/StringList.cpp (hunk 4 above)
Comment (arigliano) [reply src/lib/ngsi/StringList.cpp]: I had not tried all the tests yet. I've modified the render method, accordingly to what 'ive done in parseStringList() method, that is to add the fieldName parameter to the render() method.
Now everything seems ok and tests passed, fixing with be6a06c

File: src/lib/jsonParseV2/parseBatchQuery.cpp (hunk 3 above)
Comment (arigliano) [reply src/lib/jsonParseV2/parseBatchQuery.cpp]: Fixed in be6a06c

File: src/lib/ngsi/StringList.cpp (hunk 5 above)
Comment (arigliano) [reply src/lib/ngsi/StringList.cpp]: Fixed in be6a06c

File: doc/manuals/user/docker.md (hunk 2 above)
Comment (fgalan) [reply doc/manuals/user/docker.md]: NTC

**Comment by fgalan** [comment]
LGTM... and Merry Christmas!

**Comment by kzangeli** [comment]
LGTM and ... sorry about the delay ... (almost Happy Easter :-))

**Comment by fgalan** [comment]
After merging this PR we found some problems in our jenkins CI platform with regards functional tests. Going back to this PR, testing at its last commit point (https://github.com/telefonicaid/fiware-orion/pull/3056/commits/be6a06c09e1a693b75a1af6678d5c65cce570241), it seem the PR broke 3 tests:

Checking we are testing the right version:

```
fermin@neodeb:~/src/fiware-orion/test/functionalTest$ contextBroker --version
1.10.0-next (git version: be6a06c09e1a693b75a1af6678d5c65cce570241)
Copyright 2013-2017 Telefonica Investigacion y Desarrollo, S.A.U
Orion Context Broker is free software: you can redistribute it and/or
modify it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

Orion Context Broker is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
General Public License for more details.

Telefonica I+D
```

Failing test 1:

```
fermin@neodeb:~/src/fiware-orion/test/functionalTest$ ./testHarness.sh cases/0876_attribute_dates/get_attr_dates_with_metadata.test
mié ene 10 12:14:45 CET 2018
0001/1: get_attr_dates_with_metadata.test ............................................................................  (ERROR 9 - .out and .regexpect differ) 
0001/1: get_attr_dates_with_metadata.test ............................................................................ (intent 2) (ERROR 9 - .out and .regexpect differ) 
0001/1: get_attr_dates_with_metadata.test ............................................................................ (intent 3) (FAIL 9 - .out and .regexpect differ) 
08 seconds
Total test time: 7.85 seconds

Orion Functional Test Log File:
================================================================================
Orion Functional tests starting
Wed Jan 10 12:14:45 CET 2018

-----  Check metadata param with dates in all retrieve operations where it makes sense  -----
(get_attr_dates_with_metadata.test) output not as expected
VALIDATION ERROR: input line:
   Content-Length: 141
does not match ref line:
   Content-Length: 139


================================================================================

----------- Failing tests ------------------
  o  get_attr_dates_with_metadata.test
```

Failing test 2:

```
fermin@neodeb:~/src/fiware-orion/test/functionalTest$ ./testHarness.sh cases/1080_batch_query_context/batch_query_context_error_conditions.test
mié ene 10 12:15:12 CET 2018
0001/1: batch_query_context_error_conditions.test ....................................................................  (ERROR 9 - .out and .regexpect differ) 
0001/1: batch_query_context_error_conditions.test .................................................................... (intent 2) (ERROR 9 - .out and .regexpect differ) 
0001/1: batch_query_context_error_conditions.test .................................................................... (intent 3) (FAIL 9 - .out and .regexpect differ) 
05 seconds
Total test time: 5.40 seconds

Orion Functional Test Log File:
================================================================================
Orion Functional tests starting
Wed Jan 10 12:15:12 CET 2018

-----  V2 batch query - error conditions  -----
(batch_query_context_error_conditions.test) output not as expected
VALIDATION ERROR: input line:
   Content-Length: 83
does not match ref line:
   Content-Length: 82


================================================================================

----------- Failing tests ------------------
  o  batch_query_context_error_conditions.test
```

Failing test 3:

```
fermin@neodeb:~/src/fiware-orion/test/functionalTest$ ./testHarness.sh cases/2544_metadata_filtering_in_retrieve_operations/metadata_filtering_in_retrieve_operations.test
mié ene 10 12:15:34 CET 2018
0001/1: metadata_filtering_in_retrieve_operations.test ...............................................................  (ERROR 9 - .out and .regexpect differ) 
0001/1: metadata_filtering_in_retrieve_operations.test ............................................................... (intent 2) (ERROR 9 - .out and .regexpect differ) 
0001/1: metadata_filtering_in_retrieve_operations.test ............................................................... (intent 3) (FAIL 9 - .out and .regexpect differ) 
05 seconds
Total test time: 5.16 seconds

Orion Functional Test Log File:
================================================================================
Orion Functional tests starting
Wed Jan 10 12:15:34 CET 2018

-----  HTTP info in cached subscription on subUpdate  -----
(metadata_filtering_in_retrieve_operations.test) output not as expected
VALIDATION ERROR: input file and reference file have different line count


================================================================================

----------- Failing tests ------------------
  o  metadata_filtering_in_retrieve_operations.test
```
-----------

Given this is not the first time it occurs (see https://github.com/telefonicaid/fiware-orion/pull/3012#issuecomment-336404185). I'd like to know if there is some problem or missunderstanding on the requirement of passing a full regresion of make ft before casting the PR and ensuring that, commit after commit as review progress, this regression keeps ok. 

What do you think @arigliano ? Something to say about what could have happened?

**Comment by arigliano** [comment]
As far as I remember, all these tests passed, except for those regarding ipv6 or geo attributes (due to floating point precision, e.g. 3.14 vs 3.14000), maybe I might be wrong, in that case, I'm sorry. In any case, in future I will double check that my fork is updated with the last commit on the master before performing ftests.

**Pull request merged**